File: orders-protocol/src/main/scala/bookingtour/protocols/orders/api/PlannerExcursionItem.scala
package bookingtour.protocols.orders.api

import java.time.{LocalDate, LocalDateTime, LocalTime}

import bookingtour.protocols.core.newtypes.quantities.{Minute, Pax}
import bookingtour.protocols.core.values.Amount
import bookingtour.protocols.excursions.newTypes.{ExcursionId, ExcursionOfferId}
import bookingtour.protocols.orders.newTypes.PlannerExcursionItemId
import bookingtour.protocols.parties.api.PointUI
import derevo.cats.order
import derevo.circe.{decoder, encoder}
import derevo.derive
import cats.instances.all._
import bookingtour.protocols.core._

/**
  * © Alexey Toroshchin 2020.
  */
@derive(encoder, decoder, order)
final case class PlannerExcursionItem(
    id: PlannerExcursionItemId,
    created: LocalDateTime,
    updated: LocalDateTime,
    date: LocalDate,
    excursionId: ExcursionId,
    excursionName: String,
    accommodationPax: Pax,
    excursionOfferId: ExcursionOfferId,
    pickupPoint: PointUI,
    startTime: LocalTime,
    duration: Minute,
    total: Amount,
    clients: List[PlannerExcursionItemClient]
)

File: customer-api-spa/types/planner/PlannerExcursionItem.ts
import { Uuid } from '../basic/Uuid'
import { Amount } from '../bookings/Amount'
import { PointUI } from '../geo/PointUI'
import { DateString } from '../basic/DateString'
import { PlannerExcursionItemClient } from './PlannerExcursionItemClient'
import { parseLocalDate, parseLocalTime } from '../DateRange'

export type PlannerExcursionItemId = string

export interface PlannerExcursionItem {
  id: PlannerExcursionItemId;
  created: DateString;
  updated: DateString;
  date: DateString;
  excursionId: Uuid;
  excursionName: string;
  accommodationPax: number;
  excursionOfferId: Uuid;
  pickupPoint: PointUI;
  startTime: string;
  duration: number;
  total: Amount;
  clients: PlannerExcursionItemClient[];
}

export function sortByDateTime(l: PlannerExcursionItem, r: PlannerExcursionItem): number {
  const ld = parseLocalDate(l.date)
  const rd = parseLocalDate(r.date)
  const cmpDate = ld.valueOf() - rd.valueOf()
  if (cmpDate < 0){
    return -1
  } else if (cmpDate > 0) {
    return 1
  }
  const lt = parseLocalTime(l.startTime)
  const rt = parseLocalTime(r.startTime)
  const cmpTime = lt.valueOf() - rt.valueOf()
  if (cmpTime === 0){
    return 0
  } else if (cmpTime < 0){
    return -1
  } else {
    return 1
  }
}
